File: analytics/src/main/java/com/amazonaws/services/kinesisanalytics/operators/OffsetFutureTimestreamPoints.java
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0
package com.amazonaws.services.kinesisanalytics.operators;

import com.amazonaws.services.timestream.TimestreamPoint;
import org.apache.flink.streaming.api.functions.ProcessFunction;
import org.apache.flink.util.Collector;

import java.util.Collection;
import java.util.concurrent.TimeUnit;

public class OffsetFutureTimestreamPoints
        extends ProcessFunction<Collection<TimestreamPoint>, Collection<TimestreamPoint>> {

    private static final long TIMESTREAM_FUTURE_THRESHOLD = TimeUnit.MINUTES.toMillis(15);

    @Override
    public void processElement(Collection<TimestreamPoint> timestreamPoints, Context context,
                               Collector<Collection<TimestreamPoint>> collector) {

        timestreamPoints.stream()
                .filter(p -> pointTimestamp(p) > System.currentTimeMillis() + TIMESTREAM_FUTURE_THRESHOLD)
                .forEach(p -> {
                    p.setTime(context.timestamp());
                    p.setTimeUnit(TimeUnit.MILLISECONDS.name());
                });
        collector.collect(timestreamPoints);
    }

    private long pointTimestamp(TimestreamPoint point) {
        String timeUnit = TimeUnit.MILLISECONDS.name();
        if ("MILLISECONDS".equals(point.getTimeUnit())
                || "SECONDS".equals(point.getTimeUnit())
                || "MICROSECONDS".equals(point.getTimeUnit())
                || "NANOSECONDS".equals(point.getTimeUnit())) {
            timeUnit = point.getTimeUnit();
        }
        return TimeUnit.valueOf(timeUnit).toMillis(point.getTime());
    }
}

File: analytics-kotlin/src/main/kotlin/services/kinesisanalytics/operators/OffsetFutureTimestreamPoints.kt
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0
package services.kinesisanalytics.operators

import org.apache.flink.streaming.api.functions.ProcessFunction
import org.apache.flink.util.Collector
import services.timestream.TimestreamPoint
import java.util.concurrent.TimeUnit

class OffsetFutureTimestreamPoints : ProcessFunction<Collection<TimestreamPoint>, Collection<TimestreamPoint>>() {
    companion object {
        private val TIMESTREAM_FUTURE_THRESHOLD = TimeUnit.MINUTES.toMillis(15)
    }

    @Override
    override fun processElement(
        points: Collection<TimestreamPoint>, ctx: Context,
        out: Collector<Collection<TimestreamPoint>>
    ) {
        points.stream()
            .filter { pointTimestamp(it) > System.currentTimeMillis() + TIMESTREAM_FUTURE_THRESHOLD }
            .forEach {
                it.time = ctx.timestamp()
                it.timeUnit = TimeUnit.MILLISECONDS.name
            }
        out.collect(points)
    }

    private fun pointTimestamp(p: TimestreamPoint) =
        TimeUnit.valueOf(p.timeUnit
            .takeIf { it == "MILLISECONDS" || it == "SECONDS" || it == "MICROSECONDS" || it == "NANOSECONDS" }
            ?: TimeUnit.MILLISECONDS.name)
            .toMillis(p.time)
}
